src/NLog/Contexts/ThreadContext.cs

```diff
@@ -0,0 +1,217 @@
+        /// <returns>The key value, if defined; otherwise <c>null</c>.</returns>
+        private object TryGet(string key)
+        {
+            lock (syncRoot)
+            {
+                object o;
+                if (!this.Dict.TryGetValue(key, out o))
+                    o = null;
+
+                return o;
+            }
+        }
+
+        /// <summary>
+        /// Returns the keys in the context.
+        /// </summary>
+        public HashSet<string> Keys
+        {
+            get
+            {
+                return new HashSet<string>(this.Dict.Keys);
+            }
+        }
+
+
+        /// <summary>
+        /// Clears the content of current thread Mapped Context.
+        /// </summary>
+        public void Clear()
+        {
+            lock (syncRoot)
+            {
+                this.Dict.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Checks whether the specified key exists in current thread Mapped Context.
+        /// </summary>
+        /// <param name="key">item name.</param>
+        /// <returns>A boolean indicating whether the specified <paramref name="key"/> exists in current thread Mapped Context.</returns>
+        public bool Contains(string key)
+        {
+            lock (syncRoot)
+            {
+                return this.Dict.ContainsKey(key);
+            }
+        }
+
+        /// <summary>
+        /// Removes the specified <paramref name="key"/> from current thread Mapped Context.
+        /// </summary>
+        /// <param name="key">key name.</param>
+        public void Remove(string key)
+        {
+            lock (syncRoot)
+            {
+                this.Dict.Remove(key);
+            }
+        }
+
+
+        /// <summary>
+        /// Gets the Mapped Context key.
+        /// </summary>
+        /// <param name="key">key name.</param>
+        /// <param name="formatProvider"><see cref="IFormatProvider"/> to use when converting the key's value to a string.</param>
+        /// <returns>The value of <paramref name="key"/> as a string, if defined; otherwise <see cref="String.Empty"/>.</returns>
+        public string GetFormatted(string key, IFormatProvider formatProvider)
+        {
+            return FormatHelper.ConvertToString(this.TryGet(key), formatProvider);
+        }
+
+        /// <summary>
+        /// Gets the number of items in the collection.
+        /// </summary>
+        public int Count
+        {
+            get
+            {
+                lock (syncRoot)
+                {
+                    return this.Dict.Count;
+                }
+            }
+        }
+
+        #region [Enumerators]
+        /// <summary>
+        /// Returns an enumerator that iterates through Internal Dictionary structure
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+        {
+            var items = this.Dict;
```

Comment:
Do we need to clone `Dic` here? If yes, this line should be like this:

```
var items = new Dictionary<string, object>(this.Dict);
```
It works also ref types. My test is here: https://dotnetfiddle.net/ZNVqfL

What do you think?